src/common/history.c

```diff
@@ -1205,6 +1205,175 @@ GList *dt_history_duplicate(GList *hist)
+                              " ORDER BY num",
+                              -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+
+  while(sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    // operation
+    char *buf = (char *)sqlite3_column_text(stmt, 0);
+    if(buf) g_checksum_update(checksum, (const guchar *)buf, -1);
+    // op_params
+    buf = (char *)sqlite3_column_blob(stmt, 1);
+    int params_len = sqlite3_column_bytes(stmt, 1);
+    if(buf) g_checksum_update(checksum, (const guchar *)buf, params_len);
+    // blendop_params
+    buf = (char *)sqlite3_column_blob(stmt, 2);
+    params_len = sqlite3_column_bytes(stmt, 2);
+    if(buf) g_checksum_update(checksum, (const guchar *)buf, params_len);
+  }
+  sqlite3_finalize(stmt);
+
+  // get module order
+  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                              "SELECT version, iop_list"
+                              " FROM main.module_order"
+                              " WHERE imgid = ?1",
+                              -1, &stmt, NULL);
+  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
+  if(sqlite3_step(stmt) == SQLITE_ROW)
+  {
+    const int version = sqlite3_column_int(stmt, 0);
+    g_checksum_update(checksum, (const guchar *)&version, sizeof(version));
+    if(version == DT_IOP_ORDER_CUSTOM)
+    {
+      // iop_list
+      const char *buf = (char *)sqlite3_column_text(stmt, 1);
+      if(buf) g_checksum_update(checksum, (const guchar *)buf, -1);
+    }
+  }
+
+  const gsize checksum_len = g_checksum_type_get_length(G_CHECKSUM_MD5);
+  *hash = g_malloc(checksum_len);
+  gsize hash_len = checksum_len;
+  g_checksum_get_digest(checksum, *hash, &hash_len);
+//  gchar *hash_control = g_strdup(g_checksum_get_string(checksum));
+  g_checksum_free(checksum);
+
+//  double end = dt_get_wtime();
+//  printf("hash_history_compute img %d time %f\n", imgid, end-start);
+  return hash_len;
+}
+
+void dt_hash_history_write(const int32_t imgid, const dt_hash_history_t type)
+{
+  if(imgid == -1) return;
+  double start = dt_get_wtime();
+
+  guint8 *hash = NULL;
+  gsize hash_len = dt_hash_history_compute_from_db(imgid, &hash);
+  if(hash)
+  {
+    char *fields = NULL;
+    char *values = NULL;
+    char *conflict = NULL;
+    if(type & DT_HH_INITIAL)
+    {
+      fields = dt_util_dstrcat(fields, "%s,", "initial_hash");
+      values = g_strdup("?2,");
+      conflict = g_strdup("initial_hash=?2,");
+    }
+    if(type & DT_HH_DEFAULT)
+    {
+      fields = dt_util_dstrcat(fields, "%s,", "default_hash");
+      values = dt_util_dstrcat(values, "?2,");
+      conflict = dt_util_dstrcat(conflict, "default_hash=?2,");
+    }
+    if(type & DT_HH_CURRENT)
+    {
+      fields = dt_util_dstrcat(fields, "%s,", "current_hash");
+      values = dt_util_dstrcat(values, "?2,");
+      conflict = dt_util_dstrcat(conflict, "current_hash=?2,");
+    }
+    if(fields) fields[strlen(fields) - 1] = '\0';
```

Comment:
To remove the comas? Worth a comment I would say.

Reply:
Yes. I'll do.